File: source/com/barrybecker4/puzzle/bridge/model/MoveGenerator.java
// Copyright by Barry G. Becker, 2013. Licensed under MIT License: http://www.opensource.org/licenses/MIT
package com.barrybecker4.puzzle.bridge.model;

import scala.collection.JavaConversions;
import scala.collection.Seq;

import java.util.Arrays;
import java.util.LinkedList;
import java.util.List;

/**
 * Bridge Puzzle move generator. Generates valid next moves.
 *
 * @author Barry Becker
 */
public class MoveGenerator  {

    private Bridge board;

    /**
     * Constructor
     */
    public MoveGenerator(Bridge board) {
        this.board = board;
    }

    /**
     * Next moves are all the tiles that can slide into the current empty position.
     * @return List of all valid tile slides
     */
    public Seq<BridgeMove> generateMoves() {

        return board.isLightCrossed()?
                createMoves(board.getCrossedPeople(), false) :
                createMoves(board.getUncrossedPeople(), true);
    }

    private Seq<BridgeMove> createMoves(List<Integer> people, boolean crossing) {
        List<BridgeMove> moves = new LinkedList<>();

        int numPeople = people.size();
        moves.add(new BridgeMove(Arrays.asList(people.get(0)), crossing));

        if (numPeople > 1) {
            for (int i = 0; i < numPeople-1; i++) {
                for (int j = i + 1; j < numPeople; j++) {
                    moves.add(new BridgeMove(Arrays.asList(people.get(i), people.get(j)), crossing));
                }
                moves.add(new BridgeMove(Arrays.asList(people.get(i + 1)), crossing));
            }
        }

        // Put them in order of fastest to cross first. This speeds A*, but slows sequential a lot.
        //Collections.sort(moves);
        return JavaConversions.asScalaBuffer(moves).toSeq();
    }
}

File: source/scala/com/barrybecker4/puzzle/bridge/model/MoveGenerator.scala
// Copyright by Barry G. Becker, 2017. Licensed under MIT License: http://www.opensource.org/licenses/MIT
package com.barrybecker4.puzzle.bridge.model

import scala.collection.Seq


/**
  * Bridge Puzzle move generator. Generates valid next moves.
  *
  * @author Barry Becker
  */
class MoveGenerator(var board: Bridge) {

  /**
    * Next moves are all people that can move across the bridge..
    * @return List of all valid tile slides
    */
  def generateMoves: Seq[BridgeMove] = if (board.isLightCrossed) createMoves(board.getCrossedPeople, crossing = false)

  else createMoves(board.getUncrossedPeople, crossing = true)

  private def createMoves(people: List[Int], crossing: Boolean) = {
    var moves = List[BridgeMove]()
    val numPeople = people.size
    moves ::= new BridgeMove(List(people.head), crossing)
    if (numPeople > 1) {
      var i = 0
      while (i < numPeople - 1) {
        var j = i + 1
        while (j < numPeople) {
          moves ::= new BridgeMove(List(people(i), people(j)), crossing)
          j += 1; j - 1
        }
        moves ::= new BridgeMove(List(people(i + 1)), crossing)
        i += 1; i - 1
      }
    }
    // Put them in order of fastest to cross first. This speeds A*, but slows sequential a lot.
    // moves.sorted
    moves
  }
}
